[semsportal] initial contribution (#10100)
openhab/openhab-addons: bundles/org.openhab.binding.semsportal/README.md

# SEMSPortal Binding

> This binding can help you include statistics of your SEMS / GoodWe solar panel insallation into openHAB. It is a readonly connection that maps collected parameters to openhab channels. It provides current, day, month and total yields, as well as some income statistics if you have configured these in the SEMS portal. It requires a power station that is connected through the internet to the SEMS portal.

Reviewer: ```suggestion
This binding can help you include statistics of your SEMS / GoodWe solar panel insallation into openHAB. 
It is a read only connection that maps collected parameters to openHAB channels. 
It provides current, day, month and total yields, as well as some income statistics if you have configured these in the SEMS portal. 
It requires a power station that is connected through the internet to the SEMS portal.
```

Use openHAB not openhab, note uppercase HAB.
Also please place each sentence on a new line as it makes future changes cleaner when people do a PR.
Can you go over the entire file and make sure each sentence is a newline please?

Author: Done.

## Supported Things

> There is only one thing type supported in this binding, which is the powerStation. You need to provide the identifying UUID of the solar system in your account yourself. If you have multiple powerstations that report to the same account, you can configure the same account for each powerSation and only differ the unique id of the station in each Thing.

Reviewer: Perhaps consider a `portal` bridge thing ?!?!? Does this binding work on the kind of solar panels that have a powerStation / inverter built into each solar panel? Are there any limits to how many calls to the API/Portal you do per day?

Author: Yes I thought about a portal thing, but I have not found out yet what the request/response would be if you have more than 1 power station, typically, where to find the uniqueIDs of each station. The API has very limited to no documentation at all and errors are in Chinese. The requests are reverse-engineered from their own UI on the portal. I was planning to look into auto discovery later but the binding could be useful to people in its current form already because most people will have 1 station. I have several panels, they are all hooked up to 1 converter (station), so for most users 1 Thing should suffice. There is also no documentation on rate limits, but when you log in to the portal their own client does 1 minute updates so I thought a default of 5 and min of 1 minute should work.

Reviewer: If I understand correctly that the call has the uniqueID in it then perhaps multiple calls are needed and a bridge would not offer anything of value.
You probably already know but the API used to be publically available and a few projects are around including this Domoticz binding...
https://github.com/dylian94/domoticz-GoodWeSEMS/blob/master/plugin.py

Also I stumbled on this which may be how to do discovery..

https://github.com/DiedB/Homey-SolarPanels/issues/28#issuecomment-521042829

Author: I took the calls at the Domotics binding topic to implement a Discovery Service, so you need less parameters and more things are done for you by openHAB.

## Discovery

At this moment there is no discovery possible. Configuration of the powerStation unique id is done by hand in the Thing configuration page.

## Thing Configuration

The configuration of the thing is pretty straight forward. You need to have your power station set up in the SEMS portal, and you need to have an account that is allowed to view the power station data. You should log in at least once in the portal with this account to activate it. 

> The thing needs the username and password to connect and retreive the data. It also needs the unique id of the power station. It can be found after you log in to the portal, on the URL. use the value you find in the URL at the x-es (including the -s) in this example: https://www.semsportal.com/powerstation/powerstatussnmin/xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx

Reviewer: ```suggestion
The thing needs the username and password to connect and retreive the data.
It also needs the unique id of the power station. 
It can be found after you log in to the portal, on the URL.
Use the value you find in the URL at the x-es (including the -s) in this example: https://www.semsportal.com/powerstation/powerstatussnmin/xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
```

new line for each and a sentence started with a lowercase u.

## Channels

| channel       | type      | description                                                                                                |
| ------------- | --------- | ---------------------------------------------------------------------------------------------------------- |

> | onlineState   | Switch    | If the powerStation is currently on and reporting to the portal                                            |

Reviewer: I have a feeling this should change the ThingStatus and not be a separate channel.

Author: I think there's a difference: the openHAB Thing is still online as it has a connection with the SEMS portal. The communication with the SEMS portal is fine and we receive reports from it, but part of the report might be that the power station itself is currently shut down (because it turns off automatically when the power generation is low). In my interpretation, the Thing status indicates if openHAB is able to communicate with it, so it should remain online as long as the portal is returning answers and offline when it doesn't.

Reviewer: Please understand the following is only my opinion and that I do not have the solor panels nor will I use the binding. I feel that if you can not get current up to date data for your station, then it should go OFFLINE. However I guess that would depend on how often it does that, if it happens every time a cloud goes over and it is >10 times a day, I would probably feel differently. I am just making a suggestion, you know it better than me and will use it :) If you feel it should not change then please press the 'Resolve conversation' button.

Author: Your remarks got me thinking and I think the aspect you really want to know about the power station is if it is operational or not. Therefore I renamed the Switch channel to "operational". I personally think a Thing with an Offline state is kind of an error you need to look into. That might be because of my limited experience with openHAB, I only use a few bindings where Offline is almost equivalent to "look into the hardware or the binding configuration as there is an error". That is why I'm hesitant to use the Offline state: the station is just asleep and you do not need to do anything.

Reviewer 2: Indeed, this is exactly the purpose of the Thing status. It's fine if the Thing goes offline at night. You could add a note to the Thing status detail message, that it's in nightmode, if available.

Author: I'm sorry - a bit confused by this comment. Do you mean 1) the Thing Status should be "Offline" at night; or 2) the Thing Status should indicate "something wrong" and the solution I proposed - an "Operational" channel is ok?

Reviewer 2: First. Just for clarification, the Thing status *can* be offline at night, but it don't *need* to be offline.

It's not an error if a Thing is offline. If it's not reachable for whatever reason (inteded or unintended), it should be offline.

> | online_state   | Switch    | If the powerStation is currently on and reporting to the portal                                            |

Reviewer: Channels should use mixedCase naming convention. Example onlineState

Author: Ok I'm mainly experienced with ones that don't like zwave (switch_dimmer, alarm_motion etc). modified to camelCase.

Reviewer: Yes it is annoying when you follow merged examples and then get told differently when getting a review. Last night I looked through the dev guidelines and found broken links and feel this area needs work in the documentation as 3 of the latest binding submissions have all done this and it was only the 2.1 version guidelines that were actually clear on this at least in my opinion.

| lastUpdate    | DateTime  | Last time the powerStation sent information to the portal                                                  |
| currentOutput | Number    | The current output of the powerStation in Watt                                                             |
| todayTotal    | Number    | Todays total generation of the station in kWh                                                              |
| monthTotal    | Number    | This month's total generation of the station in kWh                                                        |
| overallTotal  | Number    | The total generation of the station since installation, in kWh                                             |
| todayIncome   | Number    | Todays income as reported by the portal, if you have configured the power rates of your energy provider    |

> | totalIncome   | Number    | The total income as reported by the portal, if you have configured the power rates of your energy provider |

Reviewer: Can you implement these with Units Of Measure, aka UOM?
You get automatic graphs drawn by OH3 if you implement it this way.
If you need a similar binding with examples to follow check out this one that is also getting reviewed that is found here:
https://github.com/Hilbrand/openhab-addons/blob/enphase/bundles/org.openhab.binding.enphase/src/main/resources/OH-INF/thing/thing-types.xml

Author: Done

## Parameters

> | parameter   | required? | description                                                                                                |

Reviewer: ```suggestion
| Parameter | Required ? | Description |
```

| ----------- |:---------:| ---------------------------------------------------------------------------------------------------------- |

> | username    | X         | Account name (emailaddress) at the SEMS portal. Account must have been used at least once to log in.       |

Reviewer: ```suggestion
| username    | X         | Account name (email address) at the SEMS portal. Account must have been used at least once to log in.       |
```

No need to separate these onto new lines for each sentence when they are in a table/grid format.

| password    | X         | Password of the SEMS portal                                                                                |

> | station     | X         | UUID of the station. See Thing Configuration which value to use here.                                      |

Reviewer: ```suggestion
| station     | X         | UUID of the station. See Thing Configuration for which value to use here.                                      |
```

You can trigger discovery in the add new Thing section of openHAB.
Select the SEMS binding and press the Scan button.
The discovered Power Stations will appear as new Things.

## Thing Configuration

The configuration of the Portal Thing (Bridge) is pretty straight forward. 
You need to have your power station set up in the SEMS portal, and you need to have an account that is allowed to view the power station data. 
You should log in at least once in the portal with this account to activate it. 
The Portal needs the username and password to connect and retrieve the data. 
You can configure the update frequency between 1 and 60 minutes. 
The default is 5 minutes.

Power Stations have no settings and will be auto discovered when you add a Portal Bridge.

## Channels

The Portal(Bridge) has no channels.
The Power Station Thing has the following channels:

| channel       | type             | description                                                                                                |
| ------------- | ---------------- | ---------------------------------------------------------------------------------------------------------- |

> | operational   | Switch           | If the powerStation is currently on and reporting to the portal                                            |

Author: removed the operational channel

> | operational   | Switch           | If the powerStation is currently on and reporting to the portal                                            |

Author: Removed the operational channel in code and comments
